### ISSUE TITLE
Question on diff_operators.py

### ISSUE THREAD

**Issue by DavidXu-JJ** [dsilvavinicius/differential_geometry_in_neural_implicits#2]
Hi, Thanks for your nice work.
I'm recently reading your code and I still have some questions about it. 
1. In principal curvature function:
https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/a755c0bb5ab0f7f9e12852502cbfb261ccabdb2a/diff_operators.py#L40-L50
I don't know why the output takes the negative operator. And the meaning changes when the above function is used.
https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/a755c0bb5ab0f7f9e12852502cbfb261ccabdb2a/diff_operators.py#L122
2. Since one principal curvature is corresponding to one principal direction.
I can get the principal direction with this function.
https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/a755c0bb5ab0f7f9e12852502cbfb261ccabdb2a/diff_operators.py#L55
And I can get the principal curvature with function I mention in the first question.
https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/a755c0bb5ab0f7f9e12852502cbfb261ccabdb2a/diff_operators.py#L40
I wonder which principal direction corresponds to the max curvature and which principal direction correspond to the min curvature.
3. It seems that hessian function receive the x with shape 2 in the last dim.
https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/a755c0bb5ab0f7f9e12852502cbfb261ccabdb2a/diff_operators.py#L166-L170
It seems that I will get a 2x2 hessian matrix with this function.
But the principal direction function use the hessian matrix like a 3x3 matrix.
https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/a755c0bb5ab0f7f9e12852502cbfb261ccabdb2a/diff_operators.py#L55-L70
This confuses me.
Thanks a lot in advance.

**Comment by tiagonovello** [comment]
Hi Jingwei Xu,

Thank you for the interest in our work.

We have revised the functions `principal_curvature` and `hessian`
Regarding point 1. That was a bug in our code, the correct output was `Kmin` and `Kmax`. We've fixed this in commit 1ff9fb2. Thank you for pointing this out!

For point 2. We have updated the output of the function `principal_directions`, which is now `dir_min` and `dir_max`. These are the principal directions associated to   `Kmin` and `Kmax`.
We have altered the script [`tools/estimate_mesh_curvatures.py`](https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/master/tools/estimate_mesh_curvatures.py) to save not only the mesh mean curvature in the quality field, but also two other versions of the mesh: one with the minimum curvature direction in place of the mesh normals, and the min curvature as a mesh quality parameter; and a second mesh with the maximum counterparts is saved as well. In the releases tab we have some pretrained models that you can test if everything is working. Just extract the [pretrained-models.tar.bz2](https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/releases/download/v0.0.1-pretrained/pretrained-models.tar.bz2) file into the project root and run the [`tools/estimate_mesh_curvatures.py`](https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/master/tools/estimate_mesh_curvatures.py). It should work. The image below shows the output of the script on the armadillo model. We used MeshLab for rendering it.

![armadillo_curvs](https://user-images.githubusercontent.com/37303476/234375283-b8c6b519-8000-4843-baff-24b1df7aa292.PNG)

As for point 3, that was a leftover documentation we had not updated (its up-to-date in the latest commit). If you input `x` with shape (B, N, 3) (B is the batch size, N being the number of points) and `y` with shape (B, N, 1) to `hessian`, then the output will be (B, N, 1, 3, 3).

If you have any questions, just comment below. If not, you may close the issue.

Regards,
Tiago and Guilherme

**Comment by DavidXu-JJ** [comment]
Hi, Tiago and Guilherme.
Thanks for your time and your quick reply! Your answer solves my questions above.
By the way, I find another potential bug when I test your new code. I hope my test result will help.
I write a test.py below:
```python

import torch
import torch.nn as nn
import diff_operators

class Ellipse(nn.Module):

    def __init__(self):
        super().__init__()


    def forward(self, x):
        
        ret = torch.pow(x, 2)
        ret[...,1] /= 4
        ret[...,2] /= 8
        ret = ret.sum(dim=-1,keepdims=True)

        return ret

x = torch.zeros((1,1,2),requires_grad=True)
x_ = torch.ones((1,1,1),requires_grad=True)
x = torch.cat([x,x_],dim=-1)
x[...,2] *= 1
s = Ellipse()
y = s(x)
print(x)
hess = diff_operators.hessian(y,x)

grad = diff_operators.gradient(y,x)

print(diff_operators.principal_curvature(y,x,grad,hess))

print(hess[0])
print(diff_operators.principal_directions(grad,hess[0]))
```
I write a simple implicit ellipse function `f(x,y,z)=x^2 + y^2/4 + z^2/8` for testing. And I get the following output:
```
tensor([[[0., 0., 1.]]], grad_fn=<CopySlices>)
(tensor([[2.0000]], grad_fn=<SubBackward0>), tensor([[8.0000]], grad_fn=<AddBackward0>))
tensor([[[[2.0000, 0.0000, 0.0000],
          [0.0000, 0.5000, 0.0000],
          [0.0000, 0.0000, 0.2500]]]], grad_fn=<SelectBackward>)
(tensor([[[1., 0., -0.]]], grad_fn=<DivBackward0>), tensor([[[-1.,  0.,  0.]]], grad_fn=<DivBackward0>))
```
As you can see, I take `(0,0,1)` as the point on the ellipse surface(at this point, the df/dz is not zero).
I get the principal direction (1,0,0) and (-1,0,0), which are not orthogonal to each other.
So I think there is still a bug in here.
After I check the paper mentioned in the code, I find that the U=W=0 hasn't been handled.
https://github.com/dsilvavinicius/differential_geometry_in_neural_implicits/blob/1ff9fb2e952e38845a274157288f6352ded1402f/diff_operators.py#L46-L49
You can check the bottom of the page 401(or the page 7 of the PDF).
http://nlpr-web.ia.ac.cn/2007papers/gjkw/gk17.pdf
I change the code a little, and it seems to get the right answer.
```python
    if U != 0 or W != 0:
        #first direction
        T1x = (-V + s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[2]]
        T1y = 2*U*grad[...,[2]]
        T1z = ( V - s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[0]] - 2*U*grad[...,[1]]
        dir_min =  torch.cat((T1x, T1y), -1)
        dir_min =  torch.cat((dir_min , T1z), -1)

        #second direction
        T2x = (-V - s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[2]]
        T2y = 2*U*grad[...,[2]]
        T2z = ( V + s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[0]] - 2*U*grad[...,[1]]
        dir_max =  torch.cat((T2x, T2y), -1)
        dir_max =  torch.cat((dir_max , T2z), -1)
    else:
        T1x = torch.zeros_like(grad[..., [0]])
        T1y = grad[..., [2]]
        T1z = - grad[..., [1]]
        dir_min =  torch.cat((T1x, T1y), -1)
        dir_min =  torch.cat((dir_min , T1z), -1)

        T2x = grad[..., [2]]
        T2y = torch.zeros_like(grad[..., [0]])
        T2z = - grad[..., [0]]
        dir_max =  torch.cat((T2x, T2y), -1)
        dir_max =  torch.cat((dir_max , T2z), -1)
```
Output:
```
tensor([[[0., 0., 1.]]], grad_fn=<CopySlices>)
(tensor([[2.0000]], grad_fn=<SubBackward0>), tensor([[8.0000]], grad_fn=<AddBackward0>))
tensor([[[[2.0000, 0.0000, 0.0000],
          [0.0000, 0.5000, 0.0000],
          [0.0000, 0.0000, 0.2500]]]], grad_fn=<SelectBackward>)
(tensor([[[0., 1., -0.]]], grad_fn=<DivBackward0>), tensor([[[1., 0., -0.]]], grad_fn=<DivBackward0>))
```
This output seems to get the right answer. The point (0,0,1) get the biggest curvature in tangent direction (1,0,0) and the minimum curvature in tangent direction (0,1,0).
Since I have already written the code, I will submit my subtle change by opening a Pull Request.
If you think this another issue I mentioned above about U=W=0 situation is correct, you can simply merge my Pull Request into your master branch. Or you can fix it by yourselves. You can check the correctness of the code by my test.py as well.
Thanks again for your time.

**Comment by DavidXu-JJ** [comment]
By the way, I wish to use your work in a U=W=0 situation. So I hope you can check this situation together with me.
Maybe there is a bug I don't notice in my pull request either.

**Comment by DavidXu-JJ** [comment]
By the way, I wish to use your work in a U=W=0 situation.
I'm grateful if you can check this pull request together with me.
Maybe there is a bug I don't notice in my pull request either.

**Comment by DavidXu-JJ** [comment]
I find an issue about the code I mention above:
```python
    if U != 0 or W != 0:
        #first direction
        T1x = (-V + s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[2]]
        T1y = 2*U*grad[...,[2]]
        T1z = ( V - s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[0]] - 2*U*grad[...,[1]]
        dir_min =  torch.cat((T1x, T1y), -1)
        dir_min =  torch.cat((dir_min , T1z), -1)

        #second direction
        T2x = (-V - s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[2]]
        T2y = 2*U*grad[...,[2]]
        T2z = ( V + s*torch.sqrt(torch.abs(V**2-4*U*W)+1e-10))*grad[...,[0]] - 2*U*grad[...,[1]]
        dir_max =  torch.cat((T2x, T2y), -1)
        dir_max =  torch.cat((dir_max , T2z), -1)
    else:
        T1x = torch.zeros_like(grad[..., [0]])
        T1y = grad[..., [2]]
        T1z = - grad[..., [1]]
        dir_min =  torch.cat((T1x, T1y), -1)
        dir_min =  torch.cat((dir_min , T1z), -1)

        T2x = grad[..., [2]]
        T2y = torch.zeros_like(grad[..., [0]])
        T2z = - grad[..., [0]]
        dir_max =  torch.cat((T2x, T2y), -1)
        dir_max =  torch.cat((dir_max , T2z), -1)
```
The U and W may be tensors and they can't use != operator.
I write another version implemented with `mask`.
https://github.com/DavidXu-JJ/differential_geometry_in_neural_implicits/blob/3a148d77137d4394cd44d7b6d339cb3397c5d897/diff_operators.py#L64-L103
This version needs to calculate the dir_min and dir_max two times for both `U!=0 or W!=0` and `U=W=0`.
I try to write an efficient implementation, but that may be hard to read for human. So I just settle down with this version.
Since I already know how to use it in my own project, my pull request is for your reference only.
You don't have to merge my pull request. It's totally fine.
It's better if you can share your opinion on the problem I mention above.

**Comment by tiagonovello** [comment]
Hi Jingwei Xu,

Thank a lot for noticing that we were missing the case U=W=0. This scenario was not common in our previous experiments.
You may have also noticed that we did not treated the case df/dz=0, however, it could be implement by permutating the coordinates. For code clarity reasons, we did not implement this case. In the future, we may do it if necessary.

We are going to accept the Pull request and merge it into our master branch.
I have tested it with your code and everything seems to continue working. Are you OK with being listed as a contributor in our README?

Regards,
Guilherme and Tiago

**Comment by DavidXu-JJ** [comment]
Hi Guilherme and Tiago,

For the df/dz=0 part, I notice this case and I personally think this may be harsh for a Neural Implicits to get a rigid 0 for df/dz(If I was wrong, please correct me). I understand your reason why not handle the df/dz=0 for clarity.

For the README part, sure I'm ok with being listed as a contributor.

At last, thanks for all of your time and your patient reading. Since all of my problem solved, I will close this issue here.

Best regards,
Jingwei
